core/src/main/java/io/grpc/internal/TransportTracer.java

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright 2017, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.internal;
+
+import io.grpc.Status;
+import io.grpc.StreamTracer;
+import java.util.concurrent.Callable;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * A class for gathering statistics about a transport. This is an experimental feature.
+ */
+public final class TransportTracer {
+  // streamsStarted happens serially, so a volatile is sufficient
+  private volatile long streamsStarted;
+  private long lastStreamCreatedTimeMsec;
+  private final AtomicLong streamsSucceeded = new AtomicLong();
```

Comment:
can you make these volatiles and use  a field updater?

Reply:
Done